### PR TITLE
WIP #321 play 2.6

### PR DESCRIPTION
**Pull request by pwyczes**
WIP #321
I was able to compile `sample project` - it runs standalone, but `clock` request needs some more work.
I also managed to compile `project code`, although I am not sure if did it correctly, since test are failing, and running tomcat7 returns page as text - need some more time (or maybe someone's help) to figure out what is going on. :)

I left some `// TODO #321` comments to be cleaned up, after successful migration. Most of them are about deleting commented code, which I left for comparison purposes.

PS
Any comments are more than welcome! :)

### CODE REVIEW

File: project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala [hunk 1]
```diff
@@ -30,21 +32,21 @@ object Play2WarServer {
 
   var playServer: Option[Play2WarServer] = None
 
-  def apply(contextPath: Option[String] = None) = {
+  def apply(contextPath: Option[String] = None): Unit = {
```
Comment (jroper) [inline project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala]: Your IDE may emit a warning about this without this change, but I would recommend not making sweeping changes like this across the code base or even just a single file when working on something that you want people to review, makes it very difficult for a reviewer to see what you've actually changed. If you want to add type annotations to every method, do it in a separate pull request.

File: project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala [hunk 2]
```diff
@@ -71,26 +75,33 @@ trait HttpServletRequestHandler extends RequestHandler {
...
-      var length = is.read(buffer)
-      while(length != -1){
-        output.write(buffer, 0, length)
-        length = is.read(buffer)
-      }
-      Enumerator(output.toByteArray) andThen Enumerator.eof
+  /** Create the source for the request body */
+  protected def convertRequestBody()(implicit mat: Materializer): Option[Source[ByteString, Any]] = {
+    getHttpRequest().getRichInputStream.map { is ⇒
+      StreamConverters.fromInputStream(() ⇒ is)
```
Comment (jroper) [inline project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala:83]: This is uses blocking IO, it's not the right approach, and it's a regression of the current behaviour, which uses the Servlet 3.1 non blocking IO APIs.

The hard work of converting Servlet 3.1 non blocking IO to reactive streams has already been done here:

https://github.com/jroper/reactive-streams-servlet

You just need to copy that code into this project (specifically these three classes: https://github.com/jroper/reactive-streams-servlet/tree/master/reactive-streams-servlet/src/main/java/org/reactivestreams/servlet), and then you can plumb the servlet input stream to Akka streams with non blocking IO.

File: project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala [hunk 3]
```diff
@@ -116,124 +127,140 @@ trait HttpServletRequestHandler extends RequestHandler {
...
-            case Success(_) =>
-              cleanup()
-              onHttpResponseComplete()
-            case Failure(ex) =>
-              Logger("play").debug(ex.toString)
-              onHttpResponseComplete()
+        val flow = body.dataStream.map { s =>
+          getHttpResponse().getRichOutputStream.foreach { os =>
+            os.write(s.toArray)
+            os.flush()
```
Comment (jroper) [inline project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala]: Same as previous comment, you're using blocking IO here, blocking on the application thread which will lead to thread exhaustion, this is a regression. Should use the ResponseSubscriber to convert the response to a Sink, and then you can feed dataStream into the response body.

**Review by jroper — CHANGES_REQUESTED** [review]
The IO needs to be updated to be non blocking, using the reactive streams adapters for Servlet 3.1 that I wrote, as it stands currently, it will cause big problems doing blocking IO on the applications executor.

What you've got currently could be used as is for Servlet 3.0 and earlier, but must not be used for Servlet 3.1.

File: project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala (hunk 1 above)
Comment (pwyczes) [reply project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala]: You are right, I started to do this in the first file and forgot to revert that. Will fix it.

File: project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala (hunk 2 above)
Comment (pwyczes) [reply project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala:83]: How about including this as a dependency to the project? I would prefer not to copy, but reuse your awesome work. :)

File: project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala (hunk 1 above)
Comment (pwyczes) [reply project-code/core/common/src/main/scala/play/core/server/servlet/Play2Server.scala]: Just fixed it. :)

File: project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala (hunk 2 above)
Comment (pwyczes) [reply project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala:83]: I was able to use reactive-streams-servlet. Can you please verify if this is correct?

File: project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala (hunk 3 above)
Comment (pwyczes) [reply project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala]: I was unable to use reactive-streams-servlet... :( 
`Sink.fromSubscriber(new ResponseSubscriber(asyncContext))` produces `Sink[ByteBuffer, NotUsed]` and I need (or at least I think I need) `Sink[ByteBuffer, Future[Any]]`.
Do you have any suggestion what to do with that?

**Comment by pwyczes** [comment]
@jroper I was able to somehow plug your reactive-streams-servlet (not everywhere, though) in play2-war. Can you, please, check in some spare time?

Also, I don't know, why in `result` from `futureResult.map { result =>` there is no `Content-Type` header, but I will have to figure it out later (at least if nobody knows out of the box... ;) ).

**Comment by jroper** [comment]
The content type and length are now properties of the HttpEntity (ie, the result body). You'll have to modify the code that looks for the content length too, assuming it's looking to the response headers to find it:

https://www.playframework.com/documentation/2.6.x/api/scala/index.html#play.api.http.HttpEntity

**Comment by pwyczes** [comment]
Thanks @jroper for your help. I appreciate it. :)

I was able to run `sbt tomcat7` and hit `http://localhost:8080/p2wsample/` - the page loads and the clock is ticking (although not every 100 millis - probably it buffers somewhere).

Sometimes I can see some logs about `non-recycled responses` and I have to figure out why they appear.
```[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] lip 27, 2018 5:07:17 PM org.apache.catalina.connector.CoyoteAdapter checkRecycled
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] INFO: Encountered a non-recycled response and recycled it forcedly.
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] org.apache.catalina.connector.CoyoteAdapter$RecycleRequiredException
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.catalina.connector.CoyoteAdapter.checkRecycled(CoyoteAdapter.java:592)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.http11.AbstractHttp11Processor.recycle(AbstractHttp11Processor.java:1814)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.http11.Http11Protocol$Http11ConnectionHandler.release(Http11Protocol.java:140)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:703)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.lang.Thread.run(Thread.java:748)
```

The next steps are to run `sbt itt-tomcat7` - but it fails at this moment and produces some NPEs like:

```[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] lip 27, 2018 5:06:57 PM org.apache.catalina.connector.CoyoteAdapter asyncDispatch
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] SEVERE: null
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer] java.lang.NullPointerException
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.tomcat.util.http.MimeHeaders.setValue(MimeHeaders.java:304)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.http11.AbstractHttp11Processor.prepareResponse(AbstractHttp11Processor.java:1485)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.http11.AbstractHttp11Processor.action(AbstractHttp11Processor.java:776)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.Response.action(Response.java:174)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.Response.sendHeaders(Response.java:348)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.catalina.connector.OutputBuffer.doFlush(OutputBuffer.java:349)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.catalina.connector.OutputBuffer.close(OutputBuffer.java:308)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.catalina.connector.Response.finishResponse(Response.java:553)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.catalina.connector.CoyoteAdapter.asyncDispatch(CoyoteAdapter.java:335)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.http11.AbstractHttp11Processor.asyncDispatch(AbstractHttp11Processor.java:1674)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:604)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
[info][org.codehaus.cargo.container.tomcat.Tomcat7xInstalledLocalContainer]     at java.lang.Thread.run(Thread.java:748)
```

and some timeouts:
```[info]   Scenario: Load page in a sub-sub directory *** FAILED *** (15 seconds, 6 milliseconds)
[info]   java.net.SocketTimeoutException: Read timed out
[info]   at java.net.SocketInputStream.socketRead0(Native Method)
[info]   at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
[info]   at java.net.SocketInputStream.read(SocketInputStream.java:171)
[info]   at java.net.SocketInputStream.read(SocketInputStream.java:141)
[info]   at org.apache.http.impl.io.AbstractSessionInputBuffer.fillBuffer(AbstractSessionInputBuffer.java:160)
[info]   at org.apache.http.impl.io.SocketInputBuffer.fillBuffer(SocketInputBuffer.java:84)
[info]   at org.apache.http.impl.io.AbstractSessionInputBuffer.readLine(AbstractSessionInputBuffer.java:273)
[info]   at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:140)
[info]   at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57)
[info]   at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:260)
```

I am trying to figure out what is happening right now, but if you @jroper (or anyone else involved in this project) have any hints, I'll take them all. ;)

**Comment by yanns** [comment]
Can you rebase on or merge `play2war:develop` to get your last changes on this branch? Thx!

**Comment by pwyczes** [comment]
done :)

**Comment by yanns** [comment]
@pwyczes which servlet version do you need?
We could maybe start only with the servlet 3.1 version and use https://github.com/jroper/reactive-streams-servlet more deeply.

**Comment by pwyczes** [comment]
We need version 3.0. 
3.1 is in further plans in this project.

**Comment by pwyczes** [comment]
Hi @yanns ,
I think, I am going to have some time to go back to this topic this (and maybe next) weekend - could you, please, find like 10 minutes to review my previous changes and maybe give me a small hint how to move forward? :)
I would really appreciate that.
No pressure, though. :)

**Comment by pwyczes** [comment]
Finally, I found some time to return to that topic.
I figured out, that those NPEs come from some concurrency issues (https://bz.apache.org/bugzilla/show_bug.cgi?id=47655#c9), since `MimeHeaders` is not thread safe.

I returned to using `play.api.libs.streams.Execution.Implicits.trampoline` (yes, I am aware of https://github.com/play2war/play2-war-plugin/issues/223) but this still did not help.

When I make one single request, like `curl -i http://localhost:8080/p2wsample/sub/subdir` then everything looks fine, but once I start `sbt itt-tomcat7` it fails randomly with unrecognized HTTP headers like `Invalid header: :21:51 GMT`, or `Invalid header: X-Permitted-Cross-Domain-Policiesferrer-Polimaster-only`. This also suggests those concurrency issues.

I have limited experience with `akka-streams`, so any suggestion where to look at is more than welcome. :)

**Comment by yanns** [comment]
Hi @pwyczes , I gave you direct access to https://github.com/play2war/play2-war-plugin.
I guess that you're the only chance for this project to survive 😉 

I can propose you the following: you can push your current PR into a new branch directly under the repo, so that we can collaboratively work on it. WDYT?

**Comment by pwyczes** [comment]
Thank you @yanns :)
I will push this PR as you suggested. :)

**Comment by pwyczes** [comment]
@yanns I merged my PR to branch `321_play_2_6`.
